### PR TITLE
Playlist relative paths

### PR DESCRIPTION
**Pull request by francescotintori**
## Description

Addes support for optional relative paths in playlists, with path conversion between windows and posix syntax.
Useful for sharing playlists between installations, on different OSes.

The modification implements new config parameter 'playlist_save_relative_paths': when enabled, and 
'rgui_browser_directory' is also set, retroarch merges the relative path name from the playlist with 'rgui_browser_directory', and then converts the final path in windows/posix syntax, depending on the local OS.

## Notes
With relative paths enabled, playlist entries are saved in this format:
```json   
{
      "path": "",
      "relative_path": "roms/Genesis/Sonic The Hedgehog (W) (REV01) [!].zip",
      "label": "Sonic The Hedgehog (W) (REV01) [!]",
      "core_path": "DETECT",
      "core_name": "DETECT",
      "crc32": "00000000|crc",
      "db_name": "Genesis.lpl"
},
```

## Related Issues

https://github.com/libretro/RetroArch/pull/10587

### CODE REVIEW

**Comment by twinaphex** [comment]
OK, a couple of things here definitely need to be redesigned -

1 - in libretro-common files, you cannot have any hard dependencies on retroarch code. That means you cannot use the settings struct. Also, it would be best to drop any references to RetroArch from libretro-common files, since it assumes a neutral environment where RetroArch isn't 'targeted' necessarily but a 'libretro frontend' is instead.

**Comment by francescotintori** [comment]
Hi @twinaphex 
you were right about references to retroarch in libretro common files, I fixed it.

**Comment by jdgleaver** [comment]
Okay, there are a few things that need tweaking here, but looks like a neat approach :)

I'm short on time at the moment - I'll try to review this fully tomorrow, and give you some pointers.

**Comment by jdgleaver** [comment]
@francescotintori Firstly, I'd just like to apologise if it seems like I'm being overly critical of your work so far. It's just that you've chosen to make fundamental changes to some of the most important routines in the codebase, so everything has to be double and triple checked. Most new contributors start with small, self-contained things that don't require this level of scrutiny. I guess you've just been unlucky...  :)

Anyway, let's start with some of the larger issues:

- twinaphex said you can't use the settings struck in `libretro-common` code. Unfortunately, the same is true inside `playlist.c`. I understand that this is seriously annoying (I've had to deal with it many times myself), but that's just the way it is... So you'll need to take `playlist_save_relative_paths` and `paths.directory_menu_content` as arguments to `playlist_init()` and cache the values there inside the `playlist_t` struct. (Note: This also means you can check the validity of `paths.directory_menu_content` in `playlist_init()`, which will save you some logic later)

- You've added relative path logic to `playlist_push()`, but this is not the only place where the path is set! e.g. There's `playlist_update()` too. Please check carefully to make sure you update all relevant functions :)

- Relative paths only work for new format playlists - you need to disable this functionality when using the old format

- Don't save `path` if `relative_path != NULL` - it just wastes space and increases file parsing time

There are also two things you haven't thought of:

- When using playlists shared between different platforms, `playlist_core_path_equal()` will not work correctly. You need to find a fix for this that doesn't affect performance (i.e. you can't search core info for this - it's too slow)

- You haven't deal with subsystem ROMs at all...

I'll make some more comments in the code itself...

File: libretro-common/file/file_path.c [hunk 1]
```diff
@@ -618,6 +629,91 @@ bool path_is_absolute(const char *path)
...
+ *  Under windows returns 'c:\games\retroarch\roms\zelda3.zip'
+ *
+ * Example[3]
+ *  @path = 'roms/zelda3.zip'
+ *  @base_content_directory = 'c:\games\retroarch\'
+ *
+ *  Under windows returns 'c:\games\retroarch\roms\zelda3.zip'
+ **/
+void path_resolve_to_local_file_system(char* buf, const char* path, const char* base_content_directory)
+{
```
Comment (jdgleaver) [inline libretro-common/file/file_path.c:676]: No, sorry - this needs to be rewritten. You absolutely cannot use the unsafe functions `strcpy()` and `strcat()`.

Please refer to the other routines in this file - they all take a `size_t len` argument, and only use safe functions.

File: playlist.c [hunk 2]
```diff
@@ -779,27 +784,45 @@ bool playlist_push(playlist_t *playlist,
...
    }
 
-   /* Get 'real' path */
    if (!string_is_empty(entry->path))
    {
+      /* Get 'real' path */
       strlcpy(real_path, entry->path, sizeof(real_path));
       playlist_resolve_path(PLAYLIST_SAVE, real_path, sizeof(real_path));
+
+      if (use_relative_path)
```
Comment (jdgleaver) [inline playlist.c:820]: `use_relative_path` is dependent upon `!string_is_empty(entry->path)` - so you're doing `!string_is_empty(entry->path)` twice unnecessarily. Move `bool use_relative_path` inside this brace.

Comment (jdgleaver) [reply playlist.c:820]: (I mean, inside the `if (!string_is_empty(entry->path))` check...)

File: playlist.c [hunk 3]
```diff
@@ -830,9 +853,10 @@ bool playlist_push(playlist_t *playlist,
    {
       struct playlist_entry tmp;
       const char *entry_path = playlist->entries[i].path;
-      bool equal_path        =
-         (string_is_empty(real_path) && string_is_empty(entry_path)) ||
-         playlist_path_equal(real_path, entry_path, fuzzy_archive_match);
+      const char* entry_relative_path = playlist->entries[i].relative_path;
+      bool equal_path =
+         (!string_is_empty(real_path) && !string_is_empty(entry_path) && playlist_path_equal(real_path, entry_path, fuzzy_archive_match)) ||
+         (!string_is_empty(relative_path) && !string_is_empty(entry_relative_path) && (strcmp(relative_path, entry_relative_path) == 0));
```
Comment (jdgleaver) [inline playlist.c:859]: Surely these `relative_path` path checks are redundant...? `real_path` and `entry_path` are all that should be needed here.

File: playlist.c [hunk 4]
```diff
@@ -1502,21 +1528,42 @@ void playlist_write_file(
...
+
          json_write_new_line(context.writer);
          json_write_space(context.writer, 6);
          JSON_Writer_WriteString(context.writer, "path",
                STRLEN_CONST("path"), JSON_UTF8);
          JSON_Writer_WriteColon(context.writer);
          json_write_space(context.writer, 1);
          JSON_Writer_WriteString(context.writer,
-               playlist->entries[i].path
-               ? playlist->entries[i].path
```
Comment (jdgleaver) [inline playlist.c:1513]: No, leave the path stuff here alone. For C89 compatibility you have to declare variables at the start of the scope, so your `char* path` will end up miles away, and the code will be unclear...

File: playlist.c [hunk 5]
```diff
@@ -2547,6 +2597,29 @@ playlist_t *playlist_init(const char *path, size_t size)
...
+
+   /* try use relative paths if enabled */
+   bool use_relative_path = (settings != NULL)
+      && settings->bools.playlist_save_relative_paths
+      && !string_is_empty(settings->paths.directory_menu_content);
+   if (use_relative_path)
+   {
+      for (i = 0; i < playlist-> size; i++)
+      {
+         struct playlist_entry* entry = playlist->entries + i;
```
Comment (jdgleaver) [inline playlist.c:2611]: Use `playlist->entries[i]`, for consistency

File: playlist.c [hunk 6]
```diff
@@ -2547,6 +2597,29 @@ playlist_t *playlist_init(const char *path, size_t size)
...
+         if (!string_is_empty(entry->relative_path))
+         {
+            if (!entry->path)
+               entry->path = (char*)malloc(PATH_MAX_LENGTH);
```
Comment (jdgleaver) [inline playlist.c:2615]: No, we can't assign a string of `PATH_MAX_LENGTH` to every playlist entry - that's an incredible waste of memory!

`strdup()` a local char array here, instead.

File: playlist.c [hunk 7]
```diff
@@ -2547,6 +2597,29 @@ playlist_t *playlist_init(const char *path, size_t size)
...
+            path_resolve_to_local_file_system(entry->path, entry->relative_path, settings->paths.directory_menu_content);
+
+            RARCH_LOG("Path '%s' resolved to '%s'\n", entry->relative_path, entry->path);
```
Comment (jdgleaver) [inline playlist.c:2618]: Don't log this - you'll just end up with thousands of lines of logging each time you load a playlist, and performance will be crippled.

**Comment by francescotintori** [comment]
Hello @jdgleaver,
I have no problems with criticism, being a SW developer myself I would be very wary of someone suddenly "breaking" things in my work :)
I admit not being very familiar with the retroarch code, so I may miss some things. 
Also, I haven't been coding in C for a while, so I'm a little rusty.
Anyway, I read most of your comments and they all seem fair points, let me work on it some more.

**Comment by francescotintori** [comment]
Ok, I fixed most things.
I reused an idea from https://github.com/libretro/RetroArch/pull/10615 to match core paths, maybe I should merge the two pull requests to reuse some code..
I still need to work on subsystem roms.

**Comment by jdgleaver** [comment]
Okay, thanks for your continued efforts :)

I'm afraid we're all busy at the moment preparing (yet another...) stable release for this weekend, and there's a lot to go through here so I won't have time to do it justice until next week.

I'd suggest *not* merging the two PRs, since that makes it harder to review and track changes. I'll also need to optimise core info searching after your other PR is merged, so best to wait on that.

In the meantime are you able to run callgrind for performance profiling? It would be good to get a handle on the performance impact of your changes. If you're running Linux, this is easy - on Windows you have to use WSL (which I'm not familiar with...). If you are able to get vallgrind/callgrind installed, I can give you some tips for using it.

**Comment by francescotintori** [comment]
Hello,
I added initial support to subsystem roms.
To be honest, I have no idea what they are, and I couldn't find any info in the docs.
So I tested the logic by manually forcing subsystem roms in an existing playlist..
Any suggestions for a real test? Which core uses subsystem roms?

**Comment by francescotintori** [comment]
I fear I destroyed my branch by merging the master branch in it by mistake.
I'll see if I can fix it or I'll make a new one

**Comment by francescotintori** [comment]
Ok forget it, I can't fix this branch.
I already made a new one, which merges fine with latest master, so I'll jist switch to it.

**Pull request closed without merging**